hc.1
.TH hc 1 "2015-12-05" "version 0.2.1" "Hieracles command manual"

.SH NAME
hc \- Command line tool for Hieracles
.SH SYNOPSIS
.B hc 
.I fqdn
.I action
.B [
.I options
.B ]

.SH DESCRIPTION
.PP
Hieracles is a command-line tool for analysis and deep examination
of Hiera parameters in a Puppet setup. It can be used to quickly
visualize, from a local puppet (typically on a developers
environment), all the Hiera params related to a specific node.
.PP
It's used internally at Gandi and its first incarnation is strongly
tied to Gandi puppet architecture. But Hieracles tends to become, in
time, a generic Hiera overlay visualization tool.
.PP
hc command takes at minimum 2 arguments:
.RS 4

.IP fqdn 10
the fqdn under which the node is known in the ENC dir

.IP action 
the action needs to be one of the following
.RS

.IP info 10
provides the farm, datacenter, country associated to the given fqdn.
An extra filter can be added to limit the list.
Use ruby regexp without the enclosing slashes.
.RS
eg. hc <fqdn> info timestamps
.RE

.IP facts 10
lists facts for a given fqdn, from puppetdb or local facts (see configuration).
An extra filter can be added to limit the list.
use ruby regexp without the enclosing slashes.
.RS
eg. hc <fqdn> facts architecture
.RE
.RS
eg. hc <fqdn> facts 'memory.*mb'
.RE

.IP files
list all files containing params affecting this fqdn 
(in more than commons).

.IP paths
list all file paths for files with params.

.IP modules
list modules included in the farm where the node is.

.IP params
list params for the node matching the fqdn.
An extra filter can be added to limit the list.
Use ruby regexp without the enclosing slashes.
.RS
eg. hc <fqdn> params postfix.*version
.RE
.RS
eg. hc <fqdn> params '^postfix'
.RE
.RS
eg. hc <fqdn> params 'version$'
.RE

.IP allparams
same as params but including the common.yaml params (huge)
Also accepts a search string
.RE
.RE

.SH OPTIONS

.TP
.PD 0
.B \-v
.TP
.PD
.B \-\-version
outputs version.

.TP
.PD 0
.B \-f \fRformat
.TP
.PD
.B \-\-format \fRformat
outputs in the selected format. Available formats are
csv, json, console, plain, yaml and rawyaml
.IP
Note that the difference between yaml and rawyaml 
is that rawyaml don't include comments with the 
provenance of each params.

.TP
.PD 0
.B \-p \fRextraparam=what,anotherparam=this
.TP
.PD
.B \-\-params \fRextraparam=what,anotherparam=this
makes possible to add additional params. 
Those params will be merged with the ones provided in the hierafile.

.TP
.PD 0
.B \-c \fRfile
.TP
.PD
.B \-\-config \fRfile
specify an alternative config file instead of 
~/.config/hieracles/config.yaml

.TP
.PD 0
.B \-h \fRfile
.TP
.PD
.B \-\-hierafile \fRfile
specify an alternative hierafile instead of the one 
stated in the configfile.

.TP
.PD 0
.B \-b \fRpath
.TP
.PD
.B \-\-basepath \fRpath
the basepath is prepended to the paths found in the configfile and the hierafile. Default is ./ but you can change this using the -b option.

.TP
.PD 0
.B \-e \fRpath
.TP
.PD
.B \-\-encdir \fRpath
encdir is where we can find the ENC definitions. It's stated in the configfile but can be overridden by the -e option.

.TP
.PD 0
.B \-y \fRfile
.TP
.PD
.B \-\-yaml \fRfile
use an arbitrary facts file that can be obtained with
.B "facter -y"
on the given node. This is useful for having a set of default fact
and test locally what would be the consequences of switching from 
one environment to another. It's also good for anticipation
of how new node, that don't have facts yet, will behave.

.TP
.PD 0
.B \-j \fRfile
.TP
.PD
.B \-\-json \fRfile
same as the 
.B "-y"
flag but in json format, which can be obtained on the node with 
.B "facter -j"
and this behave the same way. Note that the \-y has precedence on
the \-j in this context.

.TP
.PD 0
.B \-i
.TP
.PD
.B \-\-interactive
if a scope var is not defined or if the file declared in hiera config is not found, this entry is silently ignored.
.RS
This option enables the 
.B "interactive mode"
in which you are prompted to fill up for undefined scope variables encountered in the hiera config file. This behavior can be made systematic by enabling 
.I "interactive: true"
in hieracles configuration file.
.RE

.TP
.PD 0
.B \-db
if
.I "usedb: false"
is specified in the hieracles configuration file, it cancels the calls to puppetdb 
and only display information according to local data.


.TP
.PD 0
.B \-nodb
if
.I "usedb: true"
is specified in the hieracles configuration file, it forces the calls to puppetdb 
for gathering facts and node information.


.SH FILES
.I ~/.config/hieracles/config.yaml
.RS
Where sits the main hieracles configuration

.SH EXAMPLES
A typical config file would look like:
.PP
.RS
---
.RE
.RS
basepath: /home/mose/git/puppet
.RE
.RS
classpath: farm_modules/%s/manifests/init.pp
.RE
.RS
hierafile: dev/hiera-local.yaml
.RE
.RS
encpath: enc
.RE
.RS
interactive: false
.RE

.SH SEE ALSO
hiera(1), puppet(8), ppdb(1)

.SH BUGS
Please report any bug to https://github.com/Gandi/hieracles/issues

.SH AUTHORS
Copyright (c) 2015 gandi.net https://gandi.net
.LP
Hieracles is written by mose@gandi.net
.LP
https://github.com/Gandi/hieracles


CHANGELOG.md
Hieracles Changelog
=======================

### 0.3.5 - 2016-02-18
- change extraparams separator from ; to , to avoid having to quote argument
- fix merge order for extraparams to override configuration and puppetdb values

### 0.3.4 - 2016-02-12
- add a farms_counted method to registry for a ponderated list of farms

### 0.3.3 - 2016-02-11
- add some convenience methods for usage as a lib
- add a nodes_data in registry module to get all info for all nodes.

### 0.3.2 - 2015-12-30
- fix ppdb that suffered a regression

### 0.3.1 - 2015-12-28
- added some relevance to the registry class
- changed the internal way to store the nodes information
- moved Config singleton to a proper class for modularity sake

### 0.3.0 - 2015-12-16
- added json output as an option to ppdb
- added some more ppdb commands
  - ppdb resources <queries>
  - ppdb same
  - ppdb factnames
- updated manpage for ppdb with proper documentation
- prepare a queries transformation class to 
  match puppetdb weird query language with a more human one
- updated ppdb zsh completion

### 0.2.2 - 2015-12-11
- remove support for ruby 1.9.3
  because of the mime-type gem

### 0.2.1 - 2015-12-11
- add configuration variables for connection to PuppetDB
- make PuppetDB calls fill up the facts for the node queried
- add a `facts` command to display the facts, provided locally 
  or from PuppetDB directly
- make a new CLI tool named ppdb for direct calls to PuppetDB
- add optional filtering on ppdb commands
- add a man page for ppdb
- add zsh completion for ppdb

### 0.2.0 - 2015-11-24
- fix merging for unsorted arrays
- translate all paths to absolute paths,
  so that hc can be executed out of puppet dir

### 0.1.7 - 2015-11-23
- various fixes on deep_merge behavior

### 0.1.6 - 2015-11-18
- add dependency on `deep-merge` gem like hiera does
- take in account the `merge_behavior` in hiera file
- make possible to use `%{::variables}` from puppet scope
  but this is a temporary implementation, 
  only works for top level variables

### 0.1.5 - 2015-11-15
- add `interactive` new option `-i` for having 
  CLI user prompted to fill up missing scope vars
- implement interpolation adapted from the code of hiera
- fixed the single option parsing for `-i`

### 0.1.4 - 2015-11-13
- no real change, just preparing for BSD port

### 0.1.3 - 2015-11-12
- added scope (facts) interpretation for the hiera file
  (very basic interpolation, not yet hiera-lib level)

### 0.1.2 - 2015-11-12
- fix yaml more for case of nilclass in params
- fix the case when a params file is empty

### 0.1.1 - 2015-11-11
- add json format output
- add an option to load facts files in json or yaml
- fix yaml output when there is a fixnum or a float in yaml

### 0.1.0 - 2015-11-10
- display full local path in params output
- add an option to display version
- fix yaml output for true and false cases
- added a simple filter feature for yaml format,
  but it only matches like a 'start_with'

### 0.0.6 - 2015-09-21
- added doc on how to build the debian package

### 0.0.5 - 2015-09-21
- made rspec tests compat with 2.14, bundled with jessie

### 0.0.4 - 2015-09-21
- another adjustment for debian packaging

### 0.0.3 - 2015-09-21
- preparing hieracles to be packageable for debian

### 0.0.2 - 2015-09-16
- added `yaml` format including comments about where the params is defined
- the uncommented yaml is now the format `rawyaml`

### 0.0.1 - 2015-09-12
- first alpha release
- added extensive testing using rspec
- add more config variables and parameters for fitting various environments
- initial transition to Github from internal v0.0.5 Gandi code.


lib/hieracles/options/hc.rb
require 'hieracles/optparse'

module Hieracles
  module Options
    class Hc < Hieracles::Optparse

      def available_options
        {
          config: {
            has_arg: true,
            aliases: ['c', 'conf', 'config']
          },
          format: {
            has_arg: true,
            aliases: ['f', 'format']
          },
          params: {
            has_arg: true,
            aliases: ['p', 'params']
          },
          hierafile: {
            has_arg: true,
            aliases: ['h', 'hierafile']
          },
          basepath: {
            has_arg: true,
            aliases: ['b', 'basepath']
          },
          encpath: {
            has_arg: true,
            aliases: ['e', 'encpath']
          },
          version: {
            has_arg: false,
            aliases: ['v', 'version']
          },
          yaml_facts: {
            has_arg: true,
            aliases: ['y', 'yaml']
          },
          json_facts: {
            has_arg: true,
            aliases: ['j', 'json']
          },
          interactive: {
            has_arg: false,
            aliases: ['i', 'interactive']
          },
          db: {
            has_arg: false,
            aliases: ['db']
          },
          nodb: {
            has_arg: false,
            aliases: ['nodb', 'no-db', 'no']
          }
        }
      end

      def self.usage
        return <<-END

Usage: hc <fqdn> <command> [extra_args]

Available commands:
  info        provides the farm, datacenter, country
              associated to the given fqdn
              An extra param can be added for filtering
              eg. hc <fqdn> info timestamp
              eg. hc <fqdn> info farm
  facts       lists facts, either provided as a fact file
              or grabbed from puppetdb.
              An extra param can be added for filtering
              eg. hc <fqdn> facts architecture
              eg. hc <fqdn> facts 'memory.*mb'
  files       list all files containing params affecting this fqdn
              (in more than commons)
  paths       list all file paths for files with params
  modules     list modules included in the farm where the node is
  params      list params for the node matching the fqdn
              An extra filter string can be added to limit the list
              use ruby regexp without the enclosing slashes
              eg. hc <fqdn> params postfix.*version
              eg. hc <fqdn> params '^postfix'
              eg. hc <fqdn> params 'version$'
  allparams   same as params but including the common.yaml params (huge)
              Also accepts a search string

Extra args:
  -f <plain|console|csv|yaml|rawyaml|json> default console
  -p extraparam=what,anotherparam=this 
  -c <configfile>
  -h <hierafile>
  -b <basepath> default ./
  -e <encdir>
  -y <fact_file> - facts in yaml format
  -j <fact_file> - facts in json format
  -v just displays the version of Hieracles
  -i - interactive mode
  -db - query puppetdb
  -nodb - do not query puppetdb
        END
      end


    end
  end
end
    
